[v1.2] Implement Pretty Histograms

lucasferrari07 opened the pull request:
- Implement printHistogram function, that uses frequencies to print a histogram of the
text letter frequency
- Change sanitize function to return only letters
- Change itemCounter function to sort letters in ASC mode

lucasferrari07 committed:
[v1.2] Implement Pretty Histograms

- Implement printHistogram function, that uses frequencies to print a histogram of the
text letter frequency
- Change sanitize function to return only letters
- Change itemCounter function to sort letters in ASC mode

lucasferrari07 committed:
Change sanitize function

- Changed sanitize function so it doesn't use the match
and join methods, using replace instead.

lucasferrari07 committed:
Fix textalyze.test.js file lint

- Removed white line before block closing.

jfarmer commented on textalyze.js:
```diff
@@ -43,9 +43,21 @@ function stringToCharacters(text) {
+ */
+function printHistogram(frequencies) {
+  frequencies.forEach((freq, value) => {
+    const percentage = `[${freq < 0.1 ? ' ' : ''}${(freq * 100).toFixed(2)}%]`;
```
Look at [String.prototype.padStart](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart).  More broadly, any time you find yourself writing one-off code like this, double check whether there's something in the standard library that will do it (and if there is, use it).

jfarmer replied:
Er, wait, you use `padStart` below!  You could just do:

```javascript
const maxWidth = 200;
const freqStr = (freq * 100).toFixed(2).padLeft(5, ' ');
const bar = '='.repeat(freq * maxWidth);
console.log(`${value} [${freqStr}%] `${bar}`);
```

jfarmer replied:
The above code makes it absolutely clear what's going on at a glance.  That's what I'm optimizing for, here.

lucasferrari07 replied:
Of course! I totally agree with you, thanks a lot!

jfarmer commented on textalyze.js:
```diff
@@ -43,9 +43,21 @@ function stringToCharacters(text) {
+function printHistogram(frequencies) {
+  frequencies.forEach((freq, value) => {
+    const percentage = `[${freq < 0.1 ? ' ' : ''}${(freq * 100).toFixed(2)}%]`;
+    const length = ''.padStart(Math.floor(freq / 0.005), '=');
```
Here I'd use `String.prototype.repeat`, so:

```javascript
const maxWidth = 200;
'='.repeat(freq * maxWidth)
```

`String.prototype.repeat` does the flooring, i.e., `str.repeat(x) === str.repeat(Math.floor(x))` for any number `x`.

lucasferrari07 replied:
A much better way to understand, thanks!!

jfarmer commented on textalyze.js:
```diff
@@ -57,19 +69,16 @@ if (require.main === module) {
   // const program = compose(printCounts, itemCounts, stringToCharacters, sanitize);
 
-  const program = x => console.table(itemCounts(stringToCharacters(sanitize(x))));
+  const program = x => printHistogram(itemFrequencyCounter(stringToCharacters(sanitize(x))));
```
Nice!

jfarmer replied:
I wouldn't call it `program`, even those that's what I did in the example I gave.  I just wanted to show that this "program" is really a composition of several simple functions.  Give it a name like `printFileHistogram` and call `x` something more descriptive like `file` or `path` or `fileName`.

jfarmer commented on textalyze.js:
```diff
@@ -6,7 +6,7 @@ const fs = require('fs');
  */
 function itemCounts(array) {
-  return array.reduce((map, a) => {
+  return array.sort().reduce((map, a) => {
```
Make sure to test that this is returning things alphabetically.

Alternatively, I wouldn't rely on `itemCounts` to return a `Map` with alphabetically sorted keys.  The reason is that something could modify the returned `Map` and add keys in a non-alphabetical order.

I'd either:

1. Implement a type that acts like map but guarantees the keys are in order (like `set` vs `sortedSet` in some languages).
2. Not assume anything about the order of the data in the map and instead iterate over the keys alphabetically when it comes time to display them.

For (2), that would mean `printHistogram` looks more like:

```javascript
Array.from(frequencies.keys()).sort().foreach(key => {
  const val = frequencies.get(key);
  ...
});
```

lucasferrari07 replied:
It makes a lot more sense, thanks!

lucasferrari07 committed:
Implement suggested changes to PR #7

- Change program function to printFileHistogram to make it easier to
understand what it does. Also changed it's argument name to file.
- Use padStart function on frequency percentage string.
- Change the way the 'bar' is built, with string.repeat function and a
max width.

jfarmer commented on textalyze.js:
```diff
@@ -43,33 +43,46 @@ function stringToCharacters(text) {
 
 if (require.main === module) {
-  const file = process.argv.length >= 3 ? process.argv[2] : 'sample_data/tale-of-two-cities.txt';
+  const path = process.argv.length >= 3 ? process.argv[2] : 'sample_data/tale-of-two-cities.txt';
```
For a user-facing program, I wouldn't use this as a default.  Rather, I'd print a *usage* message explaining to the user how to run the program with a file and then exit with `process.exit(1)`.

jfarmer commented on textalyze.js:
```diff
@@ -43,33 +43,46 @@ function stringToCharacters(text) {
+  fs.readFile(path, (err, data) => {
     if (err) {
-      console.error(`Program failed opening '${file}', please try another one.`);
+      console.error(`Program failed opening '${path}', please try another one.`);
```
We probably want to `process.exit(1)` here so we exit with a failed error code.

jfarmer replied:
Good call on `console.error` so we log to STDERR, BTW!
